woocommerce/google-listings-and-ads
# Product statistics controller

### Changes proposed in this Pull Request:

API endpoint for #153. (#246 is the front end issue).

Adds the API endpoint `GET /mc/product-statistics` to retrieve the summary statistics for products in the Merchant Center:
```json
{
    "timestamp": 1616107264,
    "statistics": {
        "active": 0,
        "expiring": 0,
        "pending": 1,
        "disapproved": 4,
        "not_synced": 14
    }
}
```

Also adds the API endpoint `GET /mc/product-statistics/refresh` to force stats refresh (see notes).

The front end would need to be hooked up to retrieve the values.

**Notes**
- Right now the statistics are cached for hour; after that time the next request refreshes the values and the timestamp in the database.
- The first four stats (`active`, `expiring`, `pending`, and `disapproved`) come directly from the `accountstatus.get` Shopping API (summing the values for each country into each stat, and **only using the Shopping** destination (retrieved by default from the endpoint).
- For `not_synced`, uses a new `ProductRepository` method to find the products that are syncable, but aren't synced yet. (cc:@nima-karimi - does that look right?)

**Follow-ups**
- Any reason to provide the cache lifetime in the API response? To show the next refresh, for example.
- Any time the Product sync executes, it should probably reset the product statistics so they're updated the next time they're viewed (again, cc:@nima-karimi)
- Should we also have an Action Scheduler job that updates the statistics every _N_ hours? Or is refresh-on-demand enough (that is, the user will see the updated numbers the next time they visit the page with the statistics summary).

### Detailed test instructions:

1. Connect to a Merchant Center account.
2. Add a few products and wait a while for them to get disapproved (https://merchants.google.com/mc/items)
3. Use a `GET /gla/mc/product-statistics` to see the statistics as provided by the Content API and calculated against the WooCommerce products
     - Check to see everything looks more or less right
3. Use a `GET /gla/mc/product-statistics` a few more times to confirm that the response is faster, and that the `timestamp` doesn't change.
3. Use a `GET /gla/mc/product-statistics/refresh` to confirm that the `timestamp` DOES update.
4. Sync some products to the MC account and refresh the statistics, to see that `not_synced` is updated.
5. Change the `ProductStatistics::STATISTICS_LIFETIME` to `10`, and then use `GET /gla/mc/product-statistics` repeatedly to see that the response is slower and the `timestamp` is updated every 10 seconds.
